src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs

```diff
@@ -32342,9 +32284,312 @@ static void F1<T1>(T1 t1)
+                Diagnostic(ErrorCode.ERR_CircularConstraint, "U").WithArguments("U", "U").WithLocation(2, 9),
+                // (2,9): error CS0454: Circular constraint dependency involving 'U' and 'U'
+                // class B<U> where U : U?, U { }
+                Diagnostic(ErrorCode.ERR_CircularConstraint, "U").WithArguments("U", "U").WithLocation(2, 9));
+        }
+
+        [Fact]
+        public void NullableTInConstraint_04()
+        {
+            var source =
+@"class A { }
+class B
+{
+    static void F1<T>() where T : T? { }
+    static void F2<T>() where T : class, T? { }
+    static void F3<T>() where T : struct, T? { }
+    static void F4<T>() where T : A, T? { }
+    static void F5<T, U>() where U : T? { }
+    static void F6<T, U>() where T : class where U : T? { }
+    static void F7<T, U>() where T : struct where U : T? { }
+}";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics(
+                // (5,20): error CS0454: Circular constraint dependency involving 'T' and 'T'
+                //     static void F2<T>() where T : class, T? { }
+                Diagnostic(ErrorCode.ERR_CircularConstraint, "T").WithArguments("T", "T").WithLocation(5, 20),
+                // (6,43): error CS0701: 'T?' is not a valid constraint. A type used as a constraint must be an interface, a non-sealed class or a type parameter.
+                //     static void F3<T>() where T : struct, T? { }
+                Diagnostic(ErrorCode.ERR_BadBoundType, "T?").WithArguments("T?").WithLocation(6, 43),
+                // (7,20): error CS0454: Circular constraint dependency involving 'T' and 'T'
+                //     static void F4<T>() where T : A, T? { }
+                Diagnostic(ErrorCode.ERR_CircularConstraint, "T").WithArguments("T", "T").WithLocation(7, 20),
+                // (10,55): error CS0701: 'T?' is not a valid constraint. A type used as a constraint must be an interface, a non-sealed class or a type parameter.
+                //     static void F7<T, U>() where T : struct where U : T? { }
+                Diagnostic(ErrorCode.ERR_BadBoundType, "T?").WithArguments("T?").WithLocation(10, 55),
+                // (4,20): error CS0454: Circular constraint dependency involving 'T' and 'T'
+                //     static void F1<T>() where T : T? { }
+                Diagnostic(ErrorCode.ERR_CircularConstraint, "T").WithArguments("T", "T").WithLocation(4, 20));
+        }
+
+        [Fact]
+        public void NullableTInConstraint_05()
+        {
+            var source =
+@"#pragma warning disable 8321
+class C
+{
+    static void M()
+    {
+        void F1<T>() where T : T? { }
+        void F2<T>() where T : class, T? { }
+        void F3<T>() where T : struct, T? { }
+        void F4<T>() where T : A, T? { }
+        void F5<T, U>() where U : T? { }
+        void F6<T, U>() where T : class where U : T? { }
```

Comment:
>F6 [](start = 13, length = 2)

It would be good to test the instantiations on `F6` and `F7` if they aren't already covered elsewhere.